src/utils/allocators.zig

```diff
@@ -2,6 +2,251 @@ const std = @import("std");
+        pub fn append(self: *Self, n: u64) !void {
+            if (n == 0) return;
+
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            const buf = try self.memory_allocator.alloc(T, n);
+            // NOTE: we do this here so bincode serialization can work correctly
+            // otherwise, we run into undefined memory which breaks bincode
+            @memset(buf, T.DEFAULT);
+
+            try self.records.append(.{
+                .is_free = true,
+                .buf = buf,
+                .global_index = self.capacity,
+                .len = buf.len,
+                .memory_index = self.memory.items.len,
+            });
+            try self.memory.append(buf);
+            self.capacity += buf.len;
+        }
+
+        pub fn alloc(self: *Self, n: u64) !struct { []T, u64 } {
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            for (0..config.max_collapse_tries) |_| {
+                return self.allocUnsafe(n) catch |err| {
+                    switch (err) {
+                        error.CollapseFailed => {
+                            if (config.thread_safe) self.mux.unlock();
+                            defer if (config.thread_safe) self.mux.lock();
+                            std.time.sleep(std.time.ns_per_ms * config.collapse_sleep_ms);
+                            // try again -- will hit collapse again
+                            continue;
+                        },
+                        else => return err,
+                    }
+                };
+            }
+            // not enough memory to allocate and no possible recycles will be perma stuck.
+            // though its possible to recover from this, its very unlikely, so we panic
+            @panic("not enough memory, and collapse failed max times");
+        }
+
+        pub fn allocUnsafe(self: *Self, n: u64) !struct { []T, u64 } {
+            // this would never succeed
+            if (n > self.capacity) return error.AllocTooBig;
+
+            // this is used to index directly into the buffer's underlying data
+            var global_index: u64 = 0;
+            // check for a buf to recycle
+            var is_possible_to_recycle = false;
+            for (self.records.items) |*record| {
+                if (record.buf.len >= n) { // allocation is possible
+                    if (record.is_free) {
+                        record.is_free = false;
+                        const buf = record.buf[0..n]; // local copy because next line will likely change the record pointer
+                        _ = try self.tryRecycleUnusedSpaceUnsafe(record, n);
+                        return .{ buf, global_index };
+                    } else {
+                        is_possible_to_recycle = true;
+                    }
+                }
+                global_index += record.buf.len;
+            }
```

Comment:
yup - `is_possible_to_recycle` is more of a safety check than anything -- when its not possible we try to collapse to make it possible, but if that fails then we panic because the allocation will never be able to succeed